source/common/network/buffered_io_socket_handle_impl.cc

```diff
@@ -0,0 +1,228 @@
+#include "common/network/buffered_io_socket_handle_impl.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/common/platform.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/assert.h"
+#include "common/common/utility.h"
+#include "common/event/file_event_impl.h"
+#include "common/network/address_impl.h"
+
+#include "absl/container/fixed_array.h"
+#include "absl/types/optional.h"
+
+namespace Envoy {
+namespace Network {
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::close() {
+  ASSERT(!closed_);
+  if (!write_shutdown_) {
+    ASSERT(writable_peer_);
+    // Notify the peer we won't write more data. shutdown(WRITE).
+    writable_peer_->setWriteEnd();
+    // Notify the peer that we no longer accept data. shutdown(RD).
+    writable_peer_->onPeerDestroy();
+    writable_peer_->maybeSetNewData();
+    writable_peer_ = nullptr;
+    write_shutdown_ = true;
+  }
+  closed_ = true;
+  return Api::ioCallUint64ResultNoError();
+}
+
+bool BufferedIoSocketHandleImpl::isOpen() const { return !closed_; }
+
+Api::IoCallUint64Result BufferedIoSocketHandleImpl::readv(uint64_t max_length,
+                                                          Buffer::RawSlice* slices,
+                                                          uint64_t num_slice) {
+  if (owned_buffer_.length() == 0) {
```

Comment:
check it's not closed?

Reply:
My understanding is that IoHandle::readv doesn't provide the ability to return "end of stream" when the read is NON_BLOCKING.

The snippet I run at my machine when the peer wrote "abc" following the shutdown(SHUT_WR)
```
READ: n=1 data: a
READ: n=1 data: b
READ: n=1 data: c
READ: n=0 errno = 0, will try again
READ: n=0 errno = 0, will not try again
```

Reply:
add above